File: Interview Preparation Kit/Dynamic Programming/Candies.go
// https://www.hackerrank.com/challenges/candies/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"math"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'candies' function below.
 *
 * The function is expected to return a LONG_INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. INTEGER_ARRAY arr
 */
func candies(n int32, arr []int32) int64 {
	up, down := make([]int, n), make([]int, n)
	up[0] = 1
	down[n-1] = 1
	result := int64(0)

	for i := int32(1); i < n; i++ {
		if arr[i] > arr[i-1] {
			up[i] = up[i-1] + 1
		} else {
			up[i] = 1
		}
	}
	for j := n - 2; j >= 0; j-- {
		if arr[j] > arr[j+1] {
			down[j] = down[j+1] + 1
		} else {
			down[j] = 1
		}
	}
	for k := int32(0); k < n; k++ {
		result += int64(math.Max(float64(up[k]), float64(down[k])))
	}

	return result
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	nTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	n := int32(nTemp)

	var arr []int32

	for i := 0; i < int(n); i++ {
		arrItemTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
		checkError(err)
		arrItem := int32(arrItemTemp)
		arr = append(arr, arrItem)
	}

	result := candies(n, arr)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: Interview Preparation Kit/Dynamic Programming/Candies.cpp
// https://www.hackerrank.com/challenges/candies/problem
#include <bits/stdc++.h>

using namespace std;

// Complete the candies function below.
long candies(int n, vector<int> arr) {
    vector<int> up(n);
    vector<int> down(n);
    up[0] = 1;
    down[n - 1] = 1;
    long result = 0;

    for (int i = 1; i < n; i++) {
        if (arr[i] > arr[i - 1]) {
            up[i] = up[i - 1] + 1;
        }
        else {
            up[i] = 1;
        }
    }
    for (int j = n - 2; j >= 0; j--) {
        if (arr[j] > arr[j + 1]) {
            down[j] = down[j + 1] + 1;
        }
        else {
            down[j] = 1;
        }
    }
    for (int k = 0; k < n; k++) {
        result += (long) max(up[k], down[k]);
    }

    return result;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        int arr_item;
        cin >> arr_item;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');

        arr[i] = arr_item;
    }

    long result = candies(n, arr);

    fout << result << "\n";

    fout.close();

    return 0;
}
